Sources/Client.swift

```diff
@@ -0,0 +1,334 @@
+        if request.payload.method == "wc_sessionUpdate" {
+            guard let params = request.payload.params,
+                case JSONRPC_2_0.Request.Params.positional(let arrayWrapper) = params, !arrayWrapper.isEmpty,
+                case JSONRPC_2_0.ValueType.object(let sessionUpdateParams) = arrayWrapper[0],
+                let requiredApproved = sessionUpdateParams["approved"],
+                case JSONRPC_2_0.ValueType.bool(let approved) = requiredApproved else {
+                    try! send(Response(payload: JSONRPC_2_0.Response.invalidJSON, url: request.url))
+                    return
+            }
+            guard let session = communicator.session(by: request.url) else { return }
+            if !approved {
+                do {
+                    try disconnect(from: session)
+                } catch { // session already disconnected
+                    delegate.client(self, didDisconnect: session)
+                }
+            }
+        } else {
+            let payload = JSONRPC_2_0.Response.methodDoesNotExistError(id: request.payload.id)
+            try! send(Response(payload: payload, url: request.url))
+        }
+    }
+
+    override func sendDisconnectSessionRequest(for session: Session) throws {
+        let request = try! UpdateSessionRequest(url: session.url, dAppInfo: session.dAppInfo.with(approved: false))!
+        try send(request, completion: nil)
+    }
+
+    override func failedToConnect(_ url: WCURL) {
+        delegate.client(self, didFailToConnect: url)
+    }
+
+    override func didDisconnect(_ session: Session) {
+        delegate.client(self, didDisconnect: session)
+    }
+
+    /// Thread-safe collection of client reponses
+    private class Responses {
+
+        private var responses = [JSONRPC_2_0.IDType: RequestResponse]()
+        private let queue: DispatchQueue
+
+        init(queue: DispatchQueue) {
+            self.queue = queue
+        }
+
+        func add(requestID: JSONRPC_2_0.IDType, response: @escaping RequestResponse) {
+            dispatchPrecondition(condition: .notOnQueue(queue))
+            queue.sync { [unowned self] in
+                self.responses[requestID] = response
+            }
+        }
+
+        func find(requestID: JSONRPC_2_0.IDType) -> RequestResponse? {
+            var result: RequestResponse?
+            dispatchPrecondition(condition: .notOnQueue(queue))
+            queue.sync { [unowned self] in
+                result = self.responses[requestID]
+            }
+            return result
+        }
+
+        func remove(requestID: JSONRPC_2_0.IDType) {
+            dispatchPrecondition(condition: .notOnQueue(queue))
+            queue.sync { [unowned self] in
+                _ = self.responses.removeValue(forKey: requestID)
+            }
+        }
+
+    }
+
+    /// https://docs.walletconnect.org/json-rpc/ethereum#parameters-3
+    public struct Transaction: Encodable {
```

Comment:
blank line missing after this struct :trollface:

Reply:
ok